service/s3/presign_post.go

```diff
@@ -0,0 +1,408 @@
+	}
+	bucketName, ok := asS3Put.bucket()
+	if !ok {
+		return out, metadata, fmt.Errorf("requested input bucketName not found on request")
+	}
+	uploadKey := asS3Put.Key
+	if uploadKey == nil {
+		return out, metadata, fmt.Errorf("PutObject input does not have a key input")
+	}
+
+	httpReq := req.Build(ctx)
+	u := httpReq.URL.String()
+
+	signingName := awsmiddleware.GetSigningName(ctx)
+	signingRegion := awsmiddleware.GetSigningRegion(ctx)
+
+	credentials, err := s.credentialsProvider.Retrieve(ctx)
+	if err != nil {
+		return out, metadata, &v4.SigningError{
+			Err: fmt.Errorf("failed to retrieve credentials: %w", err),
+		}
+	}
+	skew := internalcontext.GetAttemptSkewContext(ctx)
+	signingTime := sdk.NowTime().Add(skew)
+	expirationTime := signingTime.Add(s.expiresIn).UTC()
+
+	fields, err := s.presigner.PresignPost(
+		credentials,
+		bucketName,
+		*uploadKey,
+		signingRegion,
+		signingName,
+		signingTime,
+		s.conditions,
+		expirationTime,
+		func(o *v4.SignerOptions) {
+			o.Logger = middleware.GetLogger(ctx)
+			o.LogSigning = s.logSigning
+		})
+	if err != nil {
+		return out, metadata, &v4.SigningError{
+			Err: fmt.Errorf("failed to sign http request, %w", err),
+		}
+	}
+
+	// Other middlewares may set default values on the URL on the path or as query params. Remove them
+	baseURL := toBaseURL(u)
+
+	out.Result = &PresignedPostRequest{
+		URL:    baseURL,
+		Values: fields,
+	}
+
+	return out, metadata, nil
+}
+
+func toBaseURL(fullURL string) string {
+	a, _ := url.Parse(fullURL)
+	return a.Scheme + "://" + a.Host
+}
+
+// Adapted from existing PresignConverter middleware
+func (c presignPostConverter) ConvertToPresignMiddleware(stack *middleware.Stack, options Options) (err error) {
+	removeStepIfPresent(stack, (*acceptencodingcust.DisableGzip)(nil).ID())
+	removeStepIfPresent(stack, (*retry.Attempt)(nil).ID())
+	removeStepIfPresent(stack, (*retry.MetricsHeader)(nil).ID())
+	stack.Deserialize.Clear()
+	stack.Build.Remove((*awsmiddleware.ClientRequestID)(nil).ID())
+
+	if err := stack.Finalize.Insert(&presignContextPolyfillMiddleware{}, "Signing", middleware.Before); err != nil {
+		return err
+	}
+	stack.Build.Remove("UserAgent")
+
+	// if no expiration is set, set one
+	expiresIn := c.Expires
+	if expiresIn == 0 {
+		expiresIn = defaultExpiresIn
+	}
+
+	pmw := newPresignPostRequestMiddleware(presignPostRequestMiddlewareOptions{
+		CredentialsProvider: options.Credentials,
+		Presigner:           c.PostPresigner,
+		LogSigning:          options.ClientLogMode.IsSigning(),
+		ExpiresIn:           expiresIn,
+		Conditions:          c.Conditions,
+	})
+	if _, err := stack.Finalize.Swap("Signing", pmw); err != nil {
+		return err
+	}
+	if err = smithyhttp.AddNoPayloadDefaultContentTypeRemover(stack); err != nil {
+		return err
+	}
+	err = presignedurlcust.AddAsIsPresigningMiddleware(stack)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func removeStepIfPresent(stack *middleware.Stack, id string) {
```

Comment:
This isn't really necessary, you can just call `Remove()` and ignore the returns (won't panic or anything)

Reply:
Done